src/main/java/com/avail/optimizer/DeadCodeAnalyzer.java

```diff
@@ -0,0 +1,251 @@
+import static java.util.Collections.nCopies;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.stream.Collectors.toCollection;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * A mechanism for determining which instructions are dead versus live.
+ */
+class DeadCodeAnalyzer
+{
+	/** The policy about which kinds of {@link L2Entity} to consider. */
+	private final DataCouplingMode dataCouplingMode;
+
+	/** The {@link L2ControlFlowGraph} to analyze. */
+	private final L2ControlFlowGraph controlFlowGraph;
+
+	/**
+	 * A {@link Map} from each {@link L2PcOperand} to the {@link Set} of
+	 * {@linkplain L2Entity entities} that might be consumed after this edge.
+	 */
+	private final Map<L2PcOperand, Set<L2Entity>> edgeNeeds = new HashMap<>();
+
+	/** The {@link L2Instruction}s that have been marked as live so far. */
+	private final Set<L2Instruction> liveInstructions = new HashSet<>();
+
+	/**
+	 * Create an instance.
+	 *
+	 * @param dataCouplingMode
+	 *        The policy about what kinds of {@link L2Entity} should be traced.
+	 * @param controlFlowGraph
+	 *        The {@link L2ControlFlowGraph} being analyzed.
+	 */
+	DeadCodeAnalyzer (
+		final DataCouplingMode dataCouplingMode,
+		final L2ControlFlowGraph controlFlowGraph)
+	{
+		this.dataCouplingMode = dataCouplingMode;
+		this.controlFlowGraph = controlFlowGraph;
+	}
+
+	/**
+	 * Calculate which operations are live, either because they have a side
+	 * effect, or they produce a value consumed (recursively) by a live
+	 * instruction.
+	 */
+	void analyzeReads ()
+	{
+		assert liveInstructions.isEmpty();
+		// First seed all edges that have forcedClampedEntities.  In theory, we
+		// could do better by determining liveness by iterating, but since the
+		// clamped entities are pretty minimal, we're not likely to eliminate a
+		// lot of dead code.
+		for (final L2BasicBlock block : controlFlowGraph.basicBlockOrder)
+		{
+			for (int i = block.predecessorEdgesCount() - 1; i >= 0; i--)
+			{
+				final L2PcOperand edge = block.predecessorEdgeAt(i);
+				if (edge.forcedClampedEntities == null)
+				{
+					continue;
+				}
+				final Set<L2Entity> needs = new HashSet<>();
+				for (final L2Entity entity : edge.forcedClampedEntities)
+				{
+					if (entity instanceof L2SemanticValue)
+					{
+						if (dataCouplingMode.considersSemanticValues())
+						{
+							needs.add(entity);
+						}
+					}
+					else if (dataCouplingMode.considersRegisters())
+					{
+						needs.add(entity);
+					}
+				}
+				edgeNeeds.put(edge, needs);
+			}
+		}
+
+		// Start with each block that has no successors, or which has only
+		// backward successor edges, collected above.
+		final Deque<L2BasicBlock> toVisit =
+			controlFlowGraph.basicBlockOrder.stream()
+				.filter(block -> block.successorEdgesCopy().stream()
+					.allMatch(edgeNeeds::containsKey))
+				.collect(toCollection(ArrayDeque::new));
+		// Visit the blocks in reverse dependency order, ignoring back-edges.
```

Comment:
It's almost true.  The back-edges and the edges containing forcedClampedEntities aren't always the same, but there's always the same number, and each mechanism breaks CFG cycles at nearly the same places.

The method starts with a loop to seed every edge that has forcedClampedEntities, then the highlighted lines 129-132 seed toVisit (which is really just a set) with every block whose successor edges are in edgeNeeds (i.e., those edges have identified what they need from ancestor instructions).  The toVisit collection of blocks is then consumed, each time scanning backward through the block's instructions to find definitions and record new usages, recording useful instructions in liveInstructions (an instruction with no side-effect and not supplying something needed downstream is dead).  The resulting needs are then placed in edgeNeeds for each of the block's predecessor edges (doing special logic for phis).  When we add a predecessor edge to edgeNeeds, we check if all of the successor edges of the predecessor block have been added already to edgeNeeds, and if so we add the predecessor block to toVisit.

This accomplishes a reverse topological traversal, accumulating all liveInstructions in a single pass.  Normally one would iterate through loops and converge to figure out a minimum fixed-point set of entities (registers and semantic values) and instructions that are really live.  The forcedClampedEntities mechanism breaks the cycles ahead of time, since our loops have minimal information carried back by the Restart mechanism.